### PR TITLE
add for loop option in tags block

### PR DESCRIPTION
**Pull request by EstyFridman1**
# User description
This feature introduces support for loops within Terraform configurations
fixes:[#496](https://github.com/bridgecrewio/yor/issues/496)

---

# Generated description

Dear maintainer, below is a concise technical summary of the changes proposed in this PR:
<p>Introduce a new feature in the <code>TerraformParser</code> class to support for loops within the tags block of Terraform configurations. This enhancement allows users to dynamically generate tags using a for loop, improving flexibility in tag management. The <code>modifyBlockTags</code> and <code>parseTagAttribute</code> functions have been updated to handle the new for loop syntax. Additionally, test cases have been added to ensure the correct parsing and handling of tags with for loops.</p>
    <table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/bridgecrewio/yor/556?tool=ast&topic=For+Loop+Support>For Loop Support</a>
        </td><td>Introduce support for for loops in the <code>TerraformParser</code> class to enhance tag management flexibility.<details><summary>Modified files (3)</summary><ul><li>src/terraform/structure/terraform_parser.go</li>
<li>tests/terraform/forLoop/main.tf</li>
<li>src/terraform/structure/terraform_parser_test.go</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>Email</th><th>Commit</th><th>Date</th></tr><tr><td>57212002+ChanochShayne...</td><td>lint-fixes-542</td><td>July 25, 2024</td></tr>
<tr><td>150965779+chanaMovshow...</td><td>Skip-Comment-Feature-530</td><td>June 30, 2024</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/bridgecrewio/yor/556?tool=ast&topic=Dependency+Updates>Dependency Updates</a>
        </td><td>Update dependencies in <code>go.mod</code> to ensure compatibility and leverage new features.<details><summary>Modified files (1)</summary><ul><li>go.mod</li></ul></details><details><summary>Latest Contributors(2)</summary><table><tr><th>Email</th><th>Commit</th><th>Date</th></tr><tr><td>47931006+RabeaZr@users...</td><td>go-git-vulnerability-483</td><td>March 07, 2024</td></tr>
<tr><td>nimrodkor@gmail.com</td><td>Extend-to-support-conf...</td><td>April 30, 2023</td></tr></table></details></td></tr></table>This pull request is reviewed by Baz. Join @EstyFridman1 and the rest of your team on <a href=https://baz.co/login>(Baz)</a>.

### CODE REVIEW

**Comment by EstyFridman1** [comment]
@ChanochShayner can you review it please?
thanks

File: src/terraform/structure/terraform_parser.go [hunk 1]
```diff
@@ -799,7 +815,38 @@ func getUncloseBracketsCount(bracketsCounters map[hclsyntax.TokenType]int) int {
 	return sum
 }
 
-func (p *TerraformParser) parseTagAttribute(tokens hclwrite.Tokens) map[string]string {
+func (p *TerraformParser) parseTagAttribute(tokens hclwrite.Tokens) (map[string]string, error) {
```
Comment (ChanochShayner) [inline src/terraform/structure/terraform_parser.go:818]: Why change this func signature to return 2 args?

File: src/terraform/structure/terraform_parser_test.go [hunk 2]
```diff
@@ -50,19 +50,31 @@ func TestTerraformParser_SkipResourceByComment(t *testing.T) {
 		assert.Empty(t, p.GetSkipResourcesByComment())
 	})
 
-	t.Run("One resource with skip comment, only that resource added to skipResourcesByComment slice", func(t *testing.T) {
```
Comment (ChanochShayner) [inline src/terraform/structure/terraform_parser_test.go:53]: Why remove this test?

File: src/terraform/structure/terraform_parser_test.go [hunk 3]
```diff
@@ -50,19 +50,31 @@ func TestTerraformParser_SkipResourceByComment(t *testing.T) {
...
+		t.Fatalf("Failed to parse file: %v", err)
+	}
+	for _, block := range parsedBlocks {
+		hclBlock := block.GetRawBlock().(*hclwrite.Block)
+		tagsAttributeName, _ := parser.getTagsAttributeName(hclBlock)
+		tagsAttribute := hclBlock.Body().GetAttribute(tagsAttributeName)
+		if tagsAttribute != nil {
+			tagsTokens := tagsAttribute.Expr().BuildTokens(hclwrite.Tokens{})
+			parsedTags, _ := parser.parseTagAttribute(tagsTokens)
+			if block.GetResourceName() == "bucket_var_tags" {
```
Comment (ChanochShayner) [inline src/terraform/structure/terraform_parser_test.go:72]: Can we test more than just this resource?

**Review by ChanochShayner — COMMENTED** [review]
Very nice!
Please fix your lint errors.

File: src/terraform/structure/terraform_parser.go (hunk 1 above)
Comment (EstyFridman1) [reply src/terraform/structure/terraform_parser.go:818]: @ChanochShayner 
the unmarshal function can return error so i think that  need to add option to return error
but if you think that it unnecessary i can remove it:
do you want to replace this:
 err := dethcl.Unmarshal([]byte(hclBytes), hclData)
to this: 
dethcl.Unmarshal([]byte(hclBytes), hclData)

File: src/terraform/structure/terraform_parser_test.go (hunk 3 above)
Comment (EstyFridman1) [reply src/terraform/structure/terraform_parser_test.go:72]: @ChanochShayner 
Yes we can test also another resource 
but we need to compare it with another expected result

File: src/terraform/structure/terraform_parser.go (hunk 1 above)
Comment (EstyFridman1) [reply src/terraform/structure/terraform_parser.go]: @ChanochShayner 
fix

File: src/terraform/structure/terraform_parser_test.go (hunk 2 above)
Comment (EstyFridman1) [reply src/terraform/structure/terraform_parser_test.go:53]: @ChanochShayner 
fix

**Comment by ChanochShayner** [comment]
@EstyFridman1 you still have some lint issues.

**Review by ChanochShayner — APPROVED** [review]

**Review by achiar99 — APPROVED** [review]

**Pull request closed without merging**